### ISSUE TITLE
False positive on type bounds within a trait (type Output: Display)

### ISSUE THREAD

**Issue by BToersche** [intellij-rust/intellij-rust#8096]
## Environment

* **IntelliJ Rust plugin version:** 0.4.157.4174-212
* **Rust toolchain version:** 1.56.1 (59eed8a2a 2021-11-01) x86_64-pc-windows-msvc
* **IDE name and version:** IntelliJ IDEA 2021.2.3 Community Edition (IC-212.5457.46)
* **Operating system:** Windows 10 10.0
* **Macro expansion engine:** new
* **Name resolution engine:** new

## Problem description
When using type bounds within a trait type, these bounds are not recognized correctly, resulting in a false positive when trying to use said type.
For instance, when the trait type is bounded to Display, the Rust Plugin will show an error when trying to use the Display functionality within the same trait. See the example below.

## Steps to reproduce
Minimal code example to see the bug in action:
```rust
pub trait TypeBoundsFalsePositive {
    type Output: Display;
    fn get_value(&self) -> Self::Output;
    /// False positive: `Self::Output` doesn't implement `Display` (required by {})
    /// However, this compiles fine (as expected).
    fn get_value_str(&self) -> String { format!("{}", self.get_value()) }
}

struct Test(i32);

impl TypeBoundsFalsePositive for Test {
    type Output = i32;
    fn get_value(&self) -> Self::Output { self.0 }
    /// No false positive here, everything is O.K.
    fn get_value_str(&self) -> String { format!("{}", self.get_value()) }
}
```
